libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/TxBodySpec.hs

```diff
@@ -0,0 +1,322 @@
+
+bodyspec ::
+  forall era fn.
+  ( EraSpecTxOut era fn
+  , EraSpecCert era fn
+  ) =>
+  UTxO era ->
+  CertsEnv era ->
+  CertState era ->
+  Specification
+    fn
+    ( ShelleyTxBody era
+    , Map (TxIn (EraCrypto era)) (TxOut era)
+    , TxIn (EraCrypto era)
+    )
+bodyspec utxo certsenv certstate =
+  constrained' $ \ [var|shelleyBody|] [var|inputUtxo|] [var|feeInput|] ->
+    match shelleyBody $
+      \ [var|inputs|] [var|outputs|] [var|certs|] [var|rewAcct|] [var|fee|] _ [var|update|] _ ->
+        [ dependsOn shelleyBody fee
+        , dependsOn shelleyBody inputs
+        , dependsOn shelleyBody outputs
+        , dependsOn feeInput inputUtxo
+        , dependsOn inputs inputUtxo
+        , dependsOn outputs inputUtxo
+        , dependsOn fee feeInput
+        , dependsOn outputs inputUtxo
+        , satisfies inputUtxo (hasSize (rangeSize 2 4))
+        , subMapSubDependsOnSuper inputUtxo (lit (unUTxO utxo))
+        , assert $ update ==. lit Nothing
+        , assert $ member_ feeInput (dom_ inputUtxo)
+        , assert $ onJust (lookup_ feeInput (lit (unUTxO utxo))) (\ [var|txout|] -> fee ==. txOutCoin_ txout)
+        , assert $ inputs ==. dom_ inputUtxo
+        , reify inputUtxo (map snd . Map.toList) $
+            \ [var|txouts|] ->
+              [ dependsOn outputs txouts
+              , assert $ (sumCoin_ @fn @era outputs) ==. sumCoin_ @fn @era txouts - fee
+              ]
+        , forAll certs $ \ [var|oneCert|] -> satisfies oneCert (txCertSpec (projectEnv certsenv) certstate)
+        , assert $ sizeOf_ certs <=. 4
+        , assert $ rewAcct ==. lit Map.empty
+        ]
+
+seqToList :: IsConwayUniv fn => HasSpec fn t => Specification fn (Data.Sequence.Internal.Seq t, [t])
+seqToList = constrained' $ \ [var|seq|] [var|list|] -> reify seq toList (\x -> list ==. x)
+
+testUTxO ::
+  forall era. (Era era, HasSpec ConwayFn (TxOut era)) => IO (Map (TxIn (EraCrypto era)) (TxOut era))
+testUTxO =
+  generate $
+    genFromSpec @ConwayFn @(Map (TxIn (EraCrypto era)) (TxOut era)) (hasSize (rangeSize 7 15))
+
+-- | Exercise the 'bodyspec'
+go ::
+  forall era.
+  ( EraSpecTxOut era ConwayFn
+  , EraSpecCert era ConwayFn
+  , HasSpec ConwayFn (Tx era)
+  ) =>
+  IO ()
+go = do
+  utxo <- UTxO <$> testUTxO @era
+  certState <-
+    generate $
+      genFromSpec @ConwayFn @(CertState era)
+        (certStateSpec @ConwayFn @era) -- (lit (AccountState (Coin 1000) (Coin 100))) (lit (EpochNo 100)))
+        -- error "STOP"
+  certsEnv <- generate $ genFromSpec @ConwayFn @(CertsEnv era) certsEnvSpec
+
+  (basic, subutxo, feeinput) <-
+    generate $ genFromSpec (bodyspec @era @ConwayFn utxo certsEnv certState)
+
+  putStrLn
+    ("Input UTxO, total " ++ show (coinBalance @era utxo) ++ ", size = " ++ show (Map.size (unUTxO utxo)))
+  putPretty "\nSubMap of Utxo" (UTxO subutxo)
+  putStrLn ("SubMap of UTxO, total Coin " ++ show (coinBalance @era (UTxO subutxo)))
+  putPretty "\nfeeInput" feeinput
+  putPretty "\nTxBody" basic
```

Comment:
Another bad name for a test that perhaps should just be an actual test.

Reply:
So all these things have been made into real tests, or removed.